### PR TITLE
Continuing cleaning a few imperative bits in lexing and interpretation loop

### PR DESCRIPTION
**Pull request by herbelin**
Hi,

This is a continuation of [b8ae2de5 and b8ae2de5~](https://github.com/coq/coq/commit/b8ae2de5be242bbd1e34ec974627c81f99b16d23) about getting rid of an extra imperative field that I overlooked in the lexer (`current_file`) and to put more structure in the vernac interpretation loop, taking also into account constructive comments from Emilio.

See the commits for explanations. In the absence of objections, my idea is to commit in a day or two since this is I think no-real-risk pure cleaning.

To go further and incidentally trying to fix [#5053](https://coq.inria.fr/bugs/show_bug.cgi?id=5053), I'm blocking in trying to have only one place for interpreting Load. For instance, why the Load in Vernac.vernac_com calls Stm.interp on the contents of the file without any problem, but if I call Stm.interp from the Load in Vernacentries.vernac_load it goes into an infinite loop or raises a Vcs_aux.Expired exception. I must confess that I'm completely lost. Wouldn't be a solution to be more functional and structured in this part of the code? How to help in this direction?

### CODE REVIEW

**Review by ejgallego — COMMENTED** [review]
> because of `eval_expr` possibly printing the vernac expression given to it

Maybe here is the key to this particular case, why does `eval_expr` would have to print the expression? Ideally you could beautify by just parsing and reprinting, Coq cannot do that but how far is from it?

**Comment by herbelin** [comment]
@ejgallego: I fully agree, it is not moral that eval_expr does some re-printing. One could do better.

**Comment by ejgallego** [comment]
IMVHO the PR looks great and it is definitively a step in the right direction to clean some nasty duplication in the codebase. Assuming accidental bugs weren't introduced it should not break anything so I can see only maintenance benefits from having this in 8.6.

**Comment by mattam82** [comment]
I think this is nice too, but why not rather wait to clean up eval_expr as
you suggested? Does anything depend on this being merged right now?

**Comment by ejgallego** [comment]
@mattam82 that is a good point but I'd have to be done. Nothing depends on this, but the reduction of global state may help with avoiding bugs in plugins (hard to say what could be affected/fixed by this).

**Comment by herbelin** [comment]
I could try to give a nicer interface to `eval_expr` but there are so many things to clean somehow more urgently.

Before this patch there were potentially some bugs in reporting location in file in case of different `entry_parse` started at the same time (e.g. by using `Load` or an `entry_parse` as in gappa) because `current_file` was not synchronized (but I did not try to find an instance).

As Emilio is implicitly saying, having too many such pending PR makes working together difficult, with a lot of time spending in merging or rebasing this or that. This is clearly not a feature but a cleaning and hence a way to potentially ease further bug resolution. I cannot exclude that a bug is introduced (as in the previous case with `fail "oops"`), but if a bug is introduced I believe it will be easier to fix on the cleaned code than if a more tricky bug is reported on the original code.

**Comment by ejgallego** [comment]
Also, does the `-just-parsing` flag serve any useful purpose? If I understand things correctly, any file using notations won't correctly parse unless you interpret the corresponding commands, right?

**Comment by herbelin** [comment]
@ejgallego: good point, I don't think that this very old option (pre-1999) is meaningful without also interpreting notations. Somehow, the vio compilation is maybe the right alternative here?

File: parsing/compat.ml4 [hunk 1]
```diff
@@ -138,20 +142,22 @@ module type LexerSig = sig
...
 module type LexerSig = sig
   include Camlp4.Sig.Lexer with module Loc = CompatLoc and type Token.t = Tok.t
-  type comments_state
-  val default_comments_state : comments_state
-  val comments_state : unit -> comments_state
-  val restore_comments_state : comments_state -> unit
+  type lexer_state
+  val init_lexer_state : string option -> lexer_state
+  val set_lexer_state : unit -> lexer_state
+  val release_lexer_state : lexer_state -> unit
```
Comment (mattam82) [inline parsing/compat.ml4]: I think there's a bug here, set and release types are inverted from the above signature

File: parsing/cLexer.ml4 [hunk 2]
```diff
@@ -354,16 +349,18 @@ let rec split_comments comacc acc pos = function
 
 let extract_comments pos = split_comments [] [] pos !comments
 
-type comments_state = int option * string * bool * ((int * int) * string) list
-let restore_comments_state (o,s,b,c) =
+type lexer_state = (int option * string * bool * ((int * int) * string) list * string option)
```
Comment (ejgallego) [inline parsing/cLexer.ml4]: Yeah just a comment would do fine.

**Review by mattam82 — APPROVED** [review]
I agree this does simplify the code and does not appear to change behavior, if the contribs pass I'm for it.

**Comment by ejgallego** [comment]
Indeed IMVHO -vio would be great to be used for tools such as the beautifier or the code analysis tools such as sercomp. I am experimenting a bit with such cleanups.

ps: I've pushed a couple of minor cleanups on top of your PR there https://github.com/ejgallego/coq/commits/v8.6%2Bextra-cleaning-state-lexer-vernac

**Comment by herbelin** [comment]
@ejgallego: about removal of do_beautify, I was thinking the same, so I added your commit to the PR.

For -just-parsing, I would prefer extra opinions. The option is de facto unusable but the concept is important. In particular, I remember Gérard Huet having regularly militated for a tool which checks that a given specification is well-formed, disregarding the existence of proofs. So, in some sense, option -quick should at the same time be put more forward for this purpose, e.g. by giving it an alias such as "-well-formed-spec", "-only-spec", though I don't find the name right now.

@mattam82: Indeed, I did not test the compilation with camlp4 and there were numerous bug. (I'm fascinated how this reviewing is useful!)

I'm going to test it on contribs.

PS: There are 2 tests in the test-suite which now fail with camlp4 (success/Notations because of the overloading IDENT/KEYWORD) and (output/Arguments_renaming.v, suspectedly in relation with the way the lexer is initialized) but not with camlp5.

**Comment by ejgallego** [comment]
@herbelin I 100% agree on the importance of a "parsing only" mode, I believe that the machinery is almost there, the first part being the `-quick` mode, and the second what @gares proposed in  https://github.com/coq/ceps/blob/master/text/001-v-file-format.md .

In fact `sercomp` supports a preliminary "check file and output a machine-readable specification" mode, however I don't believe that we are close to achieve interesting functionality yet without hacks (for instance PRs such as #204 are IMO essential)

Regarding the option, I would indeed deprecate it as it really serves no purpose other than confuse people.

**Comment by herbelin** [comment]
@mattam82: contribs ok.

I finally removed the`just_parsing` flag (it was nowhere documented).

I renamed `eval_expr` into `process_expr` to highlight that it was doing side-effects. After all, `Stm.interp` and `Vernacentries.interp` are the real `eval_expr`.

**Comment by herbelin** [comment]
Committed.

**Pull request closed without merging**